File: MySort.java
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

class MySort
{
    public static final String USE = "jar mysort.jar <input file> <output file> <number of threads>";

    private String inputFile;
    private String outputFile;
    private int numThreads;

    public MySort(String inputFile, String outputFile, int numThreads)
    {
        this.inputFile = inputFile;
        this.outputFile = outputFile;
        this.numThreads = numThreads;
    }

    public void sort()
    {
        byte[] buffer = new byte[100];
        InputStream inputStream;
        OutputStream outputStream;

        try {
            inputStream = new FileInputStream(inputFile);
            outputStream = new FileOutputStream(outputFile);
            
            while (inputStream.read(buffer) == 100) {
                outputStream.write(buffer);
            }

            inputStream.close();
            outputStream.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String args[])
    {
        if (args.length != 3) {
            System.err.println(USE);
            System.exit(1);
        }

        String inputFile = new String(args[0]);
        String outputFile = new String(args[1]);
        int numThreads = Integer.parseInt(args[2]);
        MySort mysort = new MySort(inputFile, outputFile, numThreads);

        long start, end;
        double execTime;

        start = System.nanoTime();
        mysort.sort();
        end = System.nanoTime();
        execTime = Double.valueOf(end - start).doubleValue() / 1000000000.0;

        System.out.println("input file: " + inputFile);
        System.out.println("output file: " + outputFile);
        System.out.println("number of threads: " + numThreads);
        System.out.println("execution time: " + execTime + " seconds");

        System.exit(0);
    }
}

File: mysort.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

#define USE "./mysort <input file> <output file> <number of threads>"
#define BUFFER_SIZE 100

// TODO implement external sort
void mysort(char* inputFile, char* outputFile, int numThreads)
{
    char* buffer;
    FILE* fin;
    FILE* fout;

    // Open input file
    fin = fopen(inputFile, "r");
    if (fin == NULL) {
        fprintf(stderr, "fopen(%s) failed", inputFile);
        return;
    }

    // Open output file
    fout = fopen(outputFile, "w");
    if (fout == NULL) {
        fprintf(stderr, "fopen(%s) failed", outputFile);
        return;
    }

    // Allocate memory for the buffer
    buffer = (char*) malloc(sizeof(char) * BUFFER_SIZE);

    // Read the entries line by line
    // TODO sort the entries
    while (fread(&buffer[0], sizeof(char), BUFFER_SIZE, fin) == BUFFER_SIZE) {
        fwrite(&buffer[0], sizeof(char), BUFFER_SIZE, fout);
    }

    // Clear buffer and close files
    free(buffer);
    fclose(fout);
    fclose(fin);
}

int main(int argc, char** argv) {
    char* inputFile;
    char* outputFile;
    int numThreads;
    struct timeval start, end;
    double executionTime;

    if (argc != 4) {
        fprintf(stderr, USE);
        return 1;
    }

    // Read arguments
    inputFile = argv[1];
    outputFile = argv[2];
    numThreads = atoi(argv[3]);

    // Execute sort and measure execution time
    gettimeofday(&start, NULL);
    mysort(inputFile, outputFile, numThreads);
    gettimeofday(&end, NULL);
    executionTime = ((double) end.tv_sec - (double) start.tv_sec)
            + ((double) end.tv_usec - (double) start.tv_usec) / 1000000.0;
    
    printf("input file: %s\n", inputFile);
    printf("output file: %s\n", outputFile);
    printf("number of threads: %d\n", numThreads);
    printf("execution time: %lf\n", executionTime);

    return 0;
}